File: Extensification.Tests/Byte.vb

'    Extensification  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Extensification
'
'    Extensification is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Extensification is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports NUnit.Framework
Imports Extensification.ByteExts

<TestFixture>
Public Class ByteTests

#Region "Manipulation"
    ''' <summary>
    ''' Tests byte incrementation
    ''' </summary>
    <Test>
    Public Sub TestIncrement()
        Dim ExpectedByte As Byte = 5
        Dim TargetByte As Byte = 3
        TargetByte = TargetByte.Increment(2)
        Assert.AreEqual(ExpectedByte, TargetByte)
    End Sub

    ''' <summary>
    ''' Tests signed byte incrementation
    ''' </summary>
    <Test>
    Public Sub TestIncrementSigned()
        Dim ExpectedByte As SByte = 5
        Dim TargetByte As SByte = 3
        TargetByte = TargetByte.Increment(2)
        Assert.AreEqual(ExpectedByte, TargetByte)
    End Sub

    ''' <summary>
    ''' Tests byte decrementation
    ''' </summary>
    <Test>
    Public Sub TestDecrement()
        Dim ExpectedByte As Byte = 3
        Dim TargetByte As Byte = 5
        TargetByte = TargetByte.Decrement(2)
        Assert.AreEqual(ExpectedByte, TargetByte)
    End Sub

    ''' <summary>
    ''' Tests signed byte decrementation
    ''' </summary>
    <Test>
    Public Sub TestDecrementSigned()
        Dim ExpectedByte As SByte = 3
        Dim TargetByte As SByte = 5
        TargetByte = TargetByte.Decrement(2)
        Assert.AreEqual(ExpectedByte, TargetByte)
    End Sub

    ''' <summary>
    ''' Tests byte swap
    ''' </summary>
    <Test>
    Public Sub TestSwap()
        Dim ExpectedFirstByte As Byte = 9
        Dim ExpectedSecondByte As Byte = 8
        Dim TargetFirstByte As Byte = 8
        Dim TargetSecondByte As Byte = 9
        TargetFirstByte.Swap(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub

    ''' <summary>
    ''' Tests signed byte swap
    ''' </summary>
    <Test>
    Public Sub TestSwapSigned()
        Dim ExpectedFirstByte As SByte = 9
        Dim ExpectedSecondByte As SByte = 8
        Dim TargetFirstByte As SByte = 8
        Dim TargetSecondByte As SByte = 9
        TargetFirstByte.Swap(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub

    ''' <summary>
    ''' Tests byte swap if source is larger
    ''' </summary>
    <Test>
    Public Sub TestSwapIfSourceLarger()
        Dim ExpectedFirstByte As Byte = 8
        Dim ExpectedSecondByte As Byte = 10
        Dim TargetFirstByte As Byte = 10
        Dim TargetSecondByte As Byte = 8
        TargetFirstByte.SwapIfSourceLarger(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub

    ''' <summary>
    ''' Tests signed byte swap if source is larger
    ''' </summary>
    <Test>
    Public Sub TestSwapIfSourceLargerSigned()
        Dim ExpectedFirstByte As SByte = 8
        Dim ExpectedSecondByte As SByte = 10
        Dim TargetFirstByte As SByte = 10
        Dim TargetSecondByte As SByte = 8
        TargetFirstByte.SwapIfSourceLarger(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub

    ''' <summary>
    ''' Tests byte swap if target is larger
    ''' </summary>
    <Test>
    Public Sub TestSwapIfTargetLarger()
        Dim ExpectedFirstByte As Byte = 10
        Dim ExpectedSecondByte As Byte = 8
        Dim TargetFirstByte As Byte = 8
        Dim TargetSecondByte As Byte = 10
        TargetFirstByte.SwapIfTargetLarger(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub

    ''' <summary>
    ''' Tests signed byte swap if target is larger
    ''' </summary>
    <Test>
    Public Sub TestSwapIfTargetLargerSigned()
        Dim ExpectedFirstByte As SByte = 10
        Dim ExpectedSecondByte As SByte = 8
        Dim TargetFirstByte As SByte = 8
        Dim TargetSecondByte As SByte = 10
        TargetFirstByte.SwapIfTargetLarger(TargetSecondByte)
        Assert.AreEqual(ExpectedFirstByte, TargetFirstByte)
        Assert.AreEqual(ExpectedSecondByte, TargetSecondByte)
    End Sub
#End Region

#Region "Querying"
    ''' <summary>
    ''' Tests byte digit listing
    ''' </summary>
    <Test>
    Public Sub TestListDigits()
        Dim ExpectedDigits() As Byte = {7, 5}
        Dim TargetNumber As Byte = 75
        Assert.IsTrue(ExpectedDigits.SequenceEqual(TargetNumber.ListDigits))
    End Sub

    ''' <summary>
    ''' Tests signed byte digit listing
    ''' </summary>
    <Test>
    Public Sub TestListDigitsSigned()
        Dim ExpectedDigits() As SByte = {7, 5}
        Dim TargetNumber As SByte = 75
        Assert.IsTrue(ExpectedDigits.SequenceEqual(TargetNumber.ListDigits))
    End Sub

    ''' <summary>
    ''' Tests byte Armstrong number detection
    ''' </summary>
    <Test>
    Public Sub TestIsArmstrong()
        Dim TargetNumber As Byte = 153
        Assert.IsTrue(TargetNumber.IsArmstrong)
    End Sub

    ''' <summary>
    ''' Tests signed byte Armstrong number detection
    ''' </summary>
    <Test>
    Public Sub TestIsArmstrongSigned()
        Dim TargetNumber As SByte = 1
        Assert.IsTrue(TargetNumber.IsArmstrong)
    End Sub
#End Region

End Class

File: Extensification.Tests/Byte.cs

// Extensification  Copyright (C) 2020-2021  Aptivi
// 
// This file is part of Extensification
// 
// Extensification is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Extensification is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Linq;
using Extensification.ByteExts;
using NUnit.Framework;

namespace Extensification.Tests
{

    [TestFixture]
    public class ByteTests
    {

        #region Manipulation
        /// <summary>
        /// Tests byte incrementation
        /// </summary>
        [Test]
        public void TestIncrement()
        {
            byte ExpectedByte = 5;
            byte TargetByte = 3;
            TargetByte = TargetByte.Increment(2);
            Assert.AreEqual(ExpectedByte, TargetByte);
        }

        /// <summary>
        /// Tests signed byte incrementation
        /// </summary>
        [Test]
        public void TestIncrementSigned()
        {
            sbyte ExpectedByte = 5;
            sbyte TargetByte = 3;
            TargetByte = TargetByte.Increment(2);
            Assert.AreEqual(ExpectedByte, TargetByte);
        }

        /// <summary>
        /// Tests byte decrementation
        /// </summary>
        [Test]
        public void TestDecrement()
        {
            byte ExpectedByte = 3;
            byte TargetByte = 5;
            TargetByte = TargetByte.Decrement(2);
            Assert.AreEqual(ExpectedByte, TargetByte);
        }

        /// <summary>
        /// Tests signed byte decrementation
        /// </summary>
        [Test]
        public void TestDecrementSigned()
        {
            sbyte ExpectedByte = 3;
            sbyte TargetByte = 5;
            TargetByte = TargetByte.Decrement(2);
            Assert.AreEqual(ExpectedByte, TargetByte);
        }

        /// <summary>
        /// Tests byte swap
        /// </summary>
        [Test]
        public void TestSwap()
        {
            byte ExpectedFirstByte = 9;
            byte ExpectedSecondByte = 8;
            byte TargetFirstByte = 8;
            byte TargetSecondByte = 9;
            TargetFirstByte.Swap(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }

        /// <summary>
        /// Tests signed byte swap
        /// </summary>
        [Test]
        public void TestSwapSigned()
        {
            sbyte ExpectedFirstByte = 9;
            sbyte ExpectedSecondByte = 8;
            sbyte TargetFirstByte = 8;
            sbyte TargetSecondByte = 9;
            TargetFirstByte.Swap(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }

        /// <summary>
        /// Tests byte swap if source is larger
        /// </summary>
        [Test]
        public void TestSwapIfSourceLarger()
        {
            byte ExpectedFirstByte = 8;
            byte ExpectedSecondByte = 10;
            byte TargetFirstByte = 10;
            byte TargetSecondByte = 8;
            TargetFirstByte.SwapIfSourceLarger(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }

        /// <summary>
        /// Tests signed byte swap if source is larger
        /// </summary>
        [Test]
        public void TestSwapIfSourceLargerSigned()
        {
            sbyte ExpectedFirstByte = 8;
            sbyte ExpectedSecondByte = 10;
            sbyte TargetFirstByte = 10;
            sbyte TargetSecondByte = 8;
            TargetFirstByte.SwapIfSourceLarger(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }

        /// <summary>
        /// Tests byte swap if target is larger
        /// </summary>
        [Test]
        public void TestSwapIfTargetLarger()
        {
            byte ExpectedFirstByte = 10;
            byte ExpectedSecondByte = 8;
            byte TargetFirstByte = 8;
            byte TargetSecondByte = 10;
            TargetFirstByte.SwapIfTargetLarger(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }

        /// <summary>
        /// Tests signed byte swap if target is larger
        /// </summary>
        [Test]
        public void TestSwapIfTargetLargerSigned()
        {
            sbyte ExpectedFirstByte = 10;
            sbyte ExpectedSecondByte = 8;
            sbyte TargetFirstByte = 8;
            sbyte TargetSecondByte = 10;
            TargetFirstByte.SwapIfTargetLarger(ref TargetSecondByte);
            Assert.AreEqual(ExpectedFirstByte, TargetFirstByte);
            Assert.AreEqual(ExpectedSecondByte, TargetSecondByte);
        }
        #endregion

        #region Querying
        /// <summary>
        /// Tests byte digit listing
        /// </summary>
        [Test]
        public void TestListDigits()
        {
            var ExpectedDigits = new byte[] { 7, 5 };
            byte TargetNumber = 75;
            Assert.IsTrue(ExpectedDigits.SequenceEqual(TargetNumber.ListDigits()));
        }

        /// <summary>
        /// Tests signed byte digit listing
        /// </summary>
        [Test]
        public void TestListDigitsSigned()
        {
            var ExpectedDigits = new sbyte[] { 7, 5 };
            sbyte TargetNumber = 75;
            Assert.IsTrue(ExpectedDigits.SequenceEqual(TargetNumber.ListDigits()));
        }

        /// <summary>
        /// Tests byte Armstrong number detection
        /// </summary>
        [Test]
        public void TestIsArmstrong()
        {
            byte TargetNumber = 153;
            Assert.IsTrue(TargetNumber.IsArmstrong());
        }

        /// <summary>
        /// Tests signed byte Armstrong number detection
        /// </summary>
        [Test]
        public void TestIsArmstrongSigned()
        {
            sbyte TargetNumber = 1;
            Assert.IsTrue(TargetNumber.IsArmstrong());
        }
        #endregion

    }
}